File: 1D_DP/LC_300.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.06

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // binary search based O(nlogn) solution
        int n = nums.size();
        vector<int> LIS;
        for (int i = 0; i < n; i++){
            auto it = lower_bound(LIS.begin(), LIS.end(), nums[i]);
            if (it == LIS.end()){
                LIS.push_back(nums[i]);
            }
            else{
                *it = nums[i];
            }
        }
        return LIS.size();
    }
};

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // textbook O(n^2) solution
        int n = nums.size();
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++){
            for (int j = 0; j < i; j++){
                if (nums[i] > nums[j]){
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: 1D_DP/LC_300.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.06

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # O(n^2) for DP way and O(nlogn) for binary search way
        # use binary search to find every component's to be inserted position
        def lower_bound(nums, val):
            l, r = 0, len(nums)
            while l < r:
                mid = l + (r-l)//2
                if nums[mid] >= val:
                    r = mid
                else:
                    l = mid + 1
            return l
        
        LIS = []
        for num in nums:
            pos = lower_bound(LIS, num)
            if pos == len(LIS):
                LIS.append(num)
            else:
                LIS[pos] = num
        return len(LIS)

        """
        # text book solution, now TLE for latest test case = =
        n = len(nums)
        dp = [1] * n
        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], 1 + dp[j])
        return max(dp)

        """
        